### PR TITLE
[JUJU-422] Drop some invocations of space.Subnets

### PR DESCRIPTION
**Pull request by jack-w-shaw**
As stated [here](https://github.com/juju/juju/blob/207cf067e2ee4727fb072288cb4cd6af47729c2f/state/spaces.go#L70-L71), the method space.Subnets should be phased out.

In this PR, I have removed a few invocations of this method, using space.NetworkSpace instead to get subnet information, with the intention of fully removing this method soon

## Checklist

 - [?] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [?] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [ ] Comments answer the question of why design decisions were made

## QA steps

```sh
make static-analysis
go test github.com/juju/juju/apiserver/facades/client/spaces
go test github.com/juju/juju/worker/firewaller
go test github.com/juju/juju/state --test.timeout=1000s
```

### Also

Bootstrap a juju to a provider that supports spaces, I choose aws.

#### List the current spaces after a bootstrap

```sh
$ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 172.31.32.0/20
                 252.0.0.0/12  
                 252.16.0.0/12 
                 252.32.0.0/12
```

#### Add a space

I added space with a CIDR.

```sh
$ juju add-space db 172.31.32.0/20
$ juju spaces                            
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 252.0.0.0/12  
                 252.16.0.0/12 
db     1         172.31.32.0/20
                 252.32.0.0/12 

```

#### Move a space

Move the new subnet back to the alpha space

```
juju move-to-space alpha 172.31.32.0/20
Subnet 172.31.32.0/20 moved from db to alpha
```

#### List the spaces

Once moved to another space.

```
juju spaces                                    
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 172.31.32.0/20
                 252.0.0.0/12  
                 252.16.0.0/12 
                 252.32.0.0/12 
db     1        
````

#### Move a subnet to a random space

This should fail as no space other is defined.

```sh
juju move-to-space other 172.31.16.0/20
ERROR cannot update space "other": space "other" not found
```

#### Show info about our spaces

```sh
$ juju show-space alpha
space:
  id: "0"
  name: alpha
  subnets:
  - cidr: 172.31.0.0/20
    provider-id: subnet-0815d22b08e3d82d1
    provider-network-id: vpc-0d3a33e664f1a5715
    vlan-tag: 0
    zones:
    - eu-west-2c
...
applications: []
machine-count: 0

$ juju show-space db
space:
  id: "3"
  name: db
  subnets: []
applications: []
machine-count: 0
```

#### Remove a space

```sh
$ juju remove-space db
removed space "db"
$ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 172.31.32.0/20
                 252.0.0.0/12  
                 252.16.0.0/12 
                 252.32.0.0/12 
```

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

File: state/spaces.go [hunk 1]
```diff
@@ -137,15 +137,22 @@ func (s *Space) networkSpace(subnets network.SubnetInfos) (network.SpaceInfo, er
...
-			Remove: true,
-		},
+func (s *Space) RemoveSpaceOps() ([]txn.Op, error) {
+	spaceInfo, err := s.NetworkSpace()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var ops []txn.Op
+	for _, subnet := range spaceInfo.Subnets {
+		ops = append(ops, s.st.UpdateSpaceOps(subnet.ID.String(), s.Id())...)
```
Comment (manadart) [inline state/spaces.go:147]: I'm thinking this should actually be called `UpdateSubnetSpaceOps`.

**Review by manadart — COMMENTED** [review]
Looking good. QA steps should have a real test for regression. Try something like those from the [original patch](https://github.com/juju/juju/pull/11355) for this functionality:

File: apiserver/facades/client/spaces/move_test.go [hunk 2]
```diff
@@ -11,7 +11,6 @@ import (
 	jc "github.com/juju/testing/checkers"
 	gc "gopkg.in/check.v1"
 
-	netmocks "github.com/juju/juju/apiserver/common/networkingcommon/mocks"
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply apiserver/facades/client/spaces/move_test.go:14]: Then you'll like the next PR even more!

**Review by manadart — APPROVED** [review]
Looks good. Just check if you can delete that interface altogether and regen the mocks without it.

File: apiserver/facades/client/spaces/remove.go [hunk 3]
```diff
@@ -18,45 +18,38 @@ import (
...
+	RemoveSpaceOps() ([]txn.Op, error)
 }
 
 type spaceRemoveModelOp struct {
-	space   RemoveSpace
-	subnets []Subnet
+	space RemoveSpace
 }
 
 type Subnet interface {
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply apiserver/facades/client/spaces/remove.go:28]: Especially since the method signature is actually incorrect!

**Comment by jack-w-shaw** [comment]
$$merge$$

**Comment by jack-w-shaw** [comment]
$$merge$$

**Pull request merged**